### PR TITLE
feat(graphql): add MutableTypeBatchResolver

### PR DESCRIPTION
**Pull request by NoahFournier**
This adds a MutableTypeBatchResolver utility to the GraphQL engine, allowing for native batch updates. 

## Checklist
- [X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [X] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [X] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java [hunk 1]
```diff
@@ -24,4 +24,14 @@
...
     T update(@Nonnull final String urn, @Nonnull final I input, @Nonnull final QueryContext context) throws Exception;
+    
+   /**
+    * Update many entities
+    *
+    * @param input input type
+    * @param context the {@link QueryContext} corresponding to the request.
+    */
+   default <T> T batchUpdate(@Nonnull final I[] input, @Nonnull final QueryContext context) throws Exception {
+    return null;
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java:35]: Should we return null here, or simply throw an UnsupportedOperationException if the type does not implement the function?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/MutableTypeBatchResolver.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import static com.linkedin.datahub.graphql.resolvers.ResolverUtils.*;
+
+
+/**
+ * Generic GraphQL resolver responsible for performing updates against particular types.
+ *
+ * @param <I> the generated GraphQL POJO corresponding to the input type.
+ * @param <T> the generated GraphQL POJO corresponding to the return type.
+ */
+public class MutableTypeBatchResolver<I, T> implements DataFetcher<CompletableFuture<T>> {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/MutableTypeBatchResolver.java:22]: Thank you! Do you mind adding a unit test for this one?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java (hunk 1 above)
Comment (NoahFournier) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java]: I've changed the default behaviour to throw an UnsupportedOperationException

**Comment by jjoyce0510** [comment]
Hi @NoahFournier!

First of all, thank you for the hard work so far. This PR is looking great!

I'd love to understand how we can make the end to end flow work - from GraphQL all the way to mutation.

Do you mind adding just a single example of that inside of this PR? Then we can have a followup which adds it to more GraphQL entities.

Let me know!
- John

**Comment by jjoyce0510** [comment]
@NoahFournier Pinging the thread here - are you still working on this?

**Comment by NoahFournier** [comment]
Hi @jjoyce0510
Apologies for delay on this, holidays + other tasks have pulled me away from this. I'll have an update for you very soon !

**Comment by jjoyce0510** [comment]
Hmm. I do agree that the index based matching is not really ideal from an API pov. 

Can we make an API like the following:

updateDatasets(input: BatchUpdateDatasetsInput!)

Where 

type BatchUpdateDatasetsInput {
   urn: String!
   update: DatasetUpdateInput!
}

then in the GraphQL layer we'd do something like 

input.stream().map(.... convert to correct format).collect(Collectors.toList())

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/MutableTypeBatchResolver.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Generic GraphQL resolver responsible for performing updates against particular types.
+ *
+ * @param <I> the generated GraphQL POJO corresponding to the input type.
+ * @param <T> the generated GraphQL POJO corresponding to the return type.
+ */
+public class MutableTypeBatchResolver<I, T> implements DataFetcher<CompletableFuture<List<T>>> {
+
+  private static final Logger _logger = LoggerFactory.getLogger(MutableTypeBatchResolver.class.getName());
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/MutableTypeBatchResolver.java:26]: So another option would be to just have a single <I> instead of an array.

Like BatchUpdateDatasetsInput.class as the type instead of arrayUpdateType

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java [hunk 4]
```diff
@@ -15,6 +17,10 @@
      */
     Class<I> inputClass();
 
+    default Class<I[]> arrayInputClass() throws UnsupportedOperationException {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java:20]: alternative that is being suggested above: consider changing this to batchInputClass() and just having it be a single class!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java [hunk 5]
```diff
@@ -24,4 +30,14 @@
...
      */
     T update(@Nonnull final String urn, @Nonnull final I input, @Nonnull final QueryContext context) throws Exception;
+    
+   /**
+    * Update many entities
+    *
+    * @param input input type
+    * @param context the {@link QueryContext} corresponding to the request.
+    */
+   default <T> List<T> batchUpdate(@Nonnull final String[] urns, @Nonnull final I[] input, @Nonnull final QueryContext context) throws Exception {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java:40]: If you did the above suggestion, you could do 

```
default <T> List<T> batchUpdate(@Nonnull final I updateInput, QueryContext context) throws Exception
```

File: datahub-graphql-core/src/main/resources/entity.graphql [hunk 6]
```diff
@@ -185,6 +185,11 @@ type Mutation {
     """
     updateDataset(urn: String!, input: DatasetUpdateInput!): Dataset
 
+    """
+    Update the metadata about a batch of Datasets
+    """
+    updateDatasets(urns: [String!]!, inputs: [DatasetUpdateInput!]!): [Dataset]
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/entity.graphql:191]: curious to understand what you'll be using this method for. 

Is it doing things like adding owners, adding tags, adding terms, etc? If yes, we are planning to implement batch adding metadata resolvers very soon. 

If it's about updating something else (ie description), then may this makes sense to keep!

**Comment by NoahFournier** [comment]
Hey @jjoyce0510 , made some fixes, hopefully this is more what you had in mind.

I've introduced a new Mutable type, BatchMutableType, which extends MutableType and can be implemented by entities for which you would like to enable batch updating. I've written an example for Datasets, and can extend it if this suits you. The API for batching updating datasets now looks like:

updateDatasets(input: [BatchDatasetUpdateInput!]!): [Dataset]

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java [hunk 7]
```diff
@@ -9,12 +10,11 @@
  * @param <I>: The input type corresponding to the write.
  */
 public interface MutableType<I, T> {
-
     /**
      * Returns generated GraphQL class associated with the input type
      */
-    Class<I> inputClass();
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/MutableType.java:16]: unnecessary change?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/BatchMutableType.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package com.linkedin.datahub.graphql.types;
+
+import com.linkedin.datahub.graphql.QueryContext;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+public interface BatchMutableType<I, B, T> extends MutableType<I, T> {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/BatchMutableType.java:8]: nice!

File: datahub-graphql-core/src/main/resources/entity.graphql [hunk 9]
```diff
@@ -3557,6 +3562,15 @@ input DatasetUpdateInput {
     editableProperties: DatasetEditablePropertiesUpdate
 }
 
+"""
+Arguments provided to batch update Dataset entities
+"""
+input BatchDatasetUpdateInput {
+    urn: String!
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/entity.graphql:3569]: Nit: let's add comments to every field

File: datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/mutate/MutableTypeBatchResolverTest.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+import graphql.schema.DataFetchingEnvironment;
+import org.mockito.Mockito;
+import org.testng.annotations.Test;
+
+import java.util.concurrent.CompletionException;
+
+import static com.linkedin.datahub.graphql.TestUtils.*;
+import static org.testng.Assert.*;
+
+public class MutableTypeBatchResolverTest {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/mutate/MutableTypeBatchResolverTest.java:17]: As for the success case?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataset/DatasetType.java [hunk 11]
```diff
@@ -184,6 +192,30 @@ public List<BrowsePath> browsePaths(@Nonnull String urn, @Nonnull final QueryCon
         return BrowsePathsMapper.map(result);
     }
 
+    @Override
+    public List<Dataset> batchUpdate(@Nonnull BatchDatasetUpdateInput[] input, @Nonnull QueryContext context) throws Exception {
+        final CorpuserUrn actor = CorpuserUrn.createFromString(context.getAuthentication().getActor().toUrnStr());
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataset/DatasetType.java:197]: We tend to avoid using these classes if possible these days, and instead we use the more generic Urn class

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataset/DatasetType.java [hunk 12]
```diff
@@ -184,6 +192,30 @@ public List<BrowsePath> browsePaths(@Nonnull String urn, @Nonnull final QueryCon
...
+
+        final Collection<MetadataChangeProposal> proposals = Arrays.stream(input).map(updateInput -> {
+            if (isAuthorized(updateInput.getUrn(), updateInput.getUpdate(), context)) {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataset/DatasetType.java:200]: What happens if I'm authorized to update the first dataset, but not the second?

Then we have a partial update? Getting into a partially updated state like this is dangerous - it's confusing to the end user which updates worked and which did not.

Let's instead validate authorization for each dataset in the batch, only if ALL pass authorization we apply the updte.

Comment (NoahFournier) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataset/DatasetType.java:200]: If isAuthorized fails on any of the update inputs, then the method will throw an AuthorizationException - as far as I know, this should immediately exit the batchUpdate call and return to the resolver without submitting any change proposals. I can't see any dangerous side effects within the lambda, but I'm happy to deal with this in a more explicit way if you prefer.

Comment (jjoyce0510) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataset/DatasetType.java:200]: Wait you're totally right. Misread it!

**Review by jjoyce0510 — APPROVED** [review]
Thanks for your patience & persistence here @NoahFournier.

Nice work!
